source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1094,14 +1122,19 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::connPool(
     }
   }
 
-  ConnPoolsContainer& container = parent_.host_http_conn_pool_map_[host];
-  if (!container.pools_[hash_key]) {
-    container.pools_[hash_key] = parent_.parent_.factory_.allocateConnPool(
+  ConnPoolsContainer& container = parent_.getOrAllocateHttpConnPoolsContainer(host);
+
+  // note: to simplify this, we assume that the factory is only called in the scope of this
+  // function. Otherwise, we'd need to capture a few of these variables by value.
+  auto opt_pool = container.pools_->getPool(hash_key, [&]() {
+    return parent_.parent_.factory_.allocateConnPool(
         parent_.thread_local_dispatcher_, host, priority, protocol,
         have_options ? context->downstreamConnection()->socketOptions() : nullptr);
-  }
+  });
 
-  return container.pools_[hash_key].get();
+  // currently the map cannot fail -- it does not have limits.
```

Comment:
Per other comments, let's just remove the optional stuff for now and add it back later when limits are added.